File: desktop/src/api/TpGroupe-obs-api/src/main/java/com/cegedim/next/core/tpgroupeobsapi/helpers/TraceDetailsExtractor.java
package com.cegedim.next.core.tpgroupeobsapi.helpers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

@Slf4j
@RequiredArgsConstructor
public class TraceDetailsExtractor {
    private final XmlMapper xmlMapper;

    public Map<String, String> extract(String operation, Map<String, String> tags) {
        Map<String, Function<Map<String, String>, Map<String, String>>> extractors = Map.of(
                "AbonnementPS", this::extractAbonnement,
                "IDB", this::extractSimple,
                "CLC", this::extractSimple,
                "DmdePratique", this::extractDmdePratique,
                "ConventionPS", this::extractConvention,
                "restitutionAMC", this::extractRestitution,
                "InterAMC", this::extractInterAMC,
                "DmdeSignature", this::extractSignature
        );

        try {
            return extractors.getOrDefault(operation, m -> Map.of()).apply(tags);
        } catch (Exception e) {
            log.warn("Failed to extract details for operation {}: {}", operation, e.getMessage());
            return Map.of();
        }
    }

    private JsonNode fromXml(String xml) {
        try {
            return (xml != null && !xml.isBlank()) ? xmlMapper.readTree(xml) : null;
        } catch (Exception e) {
            return null;
        }
    }

    private Map<String, String> extractAbonnement(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();
        JsonNode req = fromXml(tags.get("request.content"));
        JsonNode res = fromXml(tags.get("response.content"));

        String ackCode = res != null ? res.at("/Body/DemandeAbonPSResponse/Acquittement/AckCode").asText("") : "";
        String ackMessage = res != null ? res.at("/Body/DemandeAbonPSResponse/Acquittement/AckMessage").asText("") : "";
        String typeDemande = req != null ? req.at("/Body/DemandeAbonPSRequest/ModeAppel/TypeDemande").asText("") : "";
        String statusCode = tags.getOrDefault("status.code", "0");

        info.put("TypeDemande", typeDemande);
        info.put("AckCode", ackCode);
        info.put("AckMessage", ackMessage);
        info.put("request.num.ps", tags.get("request.num.ps"));

        boolean failed = "KO".equalsIgnoreCase(ackCode) || !"0".equals(statusCode);
        info.put("status", failed ? "FAILED" : "PASSED");

        return info;
    }

    private Map<String, String> extractSimple(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();

        info.put("request.insee.beneficiaire", tags.get("request.insee.beneficiaire"));
        info.put("request.num.amc", tags.get("request.num.amc"));
        info.put("request.num.ps", tags.get("request.num.ps"));
        info.put("request.uri", tags.get("request.uri"));

        String faultStr = tags.getOrDefault("response.fault.string", "");
        String statusCode = tags.getOrDefault("status.code", "0");


        boolean isFailure = !"Retour OK".equalsIgnoreCase(faultStr) || !"0".equals(statusCode);
        info.put("status", isFailure ? "FAILED" : "PASSED");

        if (isFailure) {
            JsonNode res = fromXml(tags.get("response.content"));
            if (res != null) {
                String reason = res.at("/Body/Fault/Reason/Text/__text").asText(null);
                if (reason != null && !reason.isBlank()) {
                    info.put("failure.reason", reason);
                    info.put("response.fault.string", faultStr);
                }
            }
        }

        return info;
    }

    private Map<String, String> extractDmdePratique(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();
        JsonNode res = fromXml(tags.get("response.content"));

        String ackCode = res != null ? res.at("/Body/DonneesPSResponse/AckCode").asText("") : "";
        String ackMessage = res != null ? res.at("/Body/DonneesPSResponse/AckMessage").asText("") : "";
        String statusCode = tags.getOrDefault("status.code", "0");

        info.put("request.num.ps", tags.get("request.num.ps"));
        info.put("request.uri", tags.get("request.uri"));
        info.put("AckCode", ackCode);
        info.put("AckMessage", ackMessage);

        boolean failed = "KO".equalsIgnoreCase(ackCode) || !"0".equals(statusCode);
        info.put("status", failed ? "FAILED" : "PASSED");

        return info;
    }

    private Map<String, String> extractConvention(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();

        String faultString = tags.getOrDefault("response.fault.string", "");
        String statusCode = tags.getOrDefault("status.code", "0");

        boolean failed = !"Retour OK".equalsIgnoreCase(faultString) || !"0".equals(statusCode);

        info.put("request.num.amc", tags.get("request.num.amc"));
        info.put("request.num.ps", tags.get("request.num.ps"));
        info.put("status", failed ? "FAILED" : "PASSED");

        if (failed) {
            info.put("response.fault.string", faultString);
        }

        return info;
    }

    private Map<String, String> extractRestitution(Map<String, String> tags) {

        Map<String, String> info = new HashMap<>();

        String faultString = tags.getOrDefault("response.fault.string", "");
        String faultCode = tags.getOrDefault("response.fault.code", "");
        String statusCode = tags.getOrDefault("status.code", "0");

        boolean failed = !"Retour OK".equalsIgnoreCase(faultString) || !"0".equals(statusCode);

        info.put("request.num.ps", tags.get("request.num.ps"));
        info.put("status", failed ? "FAILED" : "PASSED");

        if (failed) {
            info.put("response.fault.string", faultString);
            info.put("response.fault.code", faultCode);
        }

        return info;
    }

    private Map<String, String> extractInterAMC(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();
        JsonNode res = fromXml(tags.get("response.content"));

        String codeRetour = "";
        String message = "";
        boolean failed = false;

        if (res != null) {
            codeRetour = res.at("/Body/searchResponse/searchResult/etat_demande/code_retour").asText("");
            message = res.at("/Body/searchResponse/searchResult/etat_demande/message").asText("");
            failed = !"1".equals(codeRetour);
        }

        info.put("status", failed ? "FAILED" : "PASSED");
        if (failed) {
            info.put("code_retour", codeRetour);
            info.put("message", message);
        }

        return info;
    }



    private Map<String, String> extractSignature(Map<String, String> tags) {
        Map<String, String> info = new HashMap<>();
        JsonNode res = fromXml(tags.get("response.content"));

        String faultString = tags.getOrDefault("response.fault.string", "");
        String statusCode = tags.getOrDefault("status.code", "0");

        boolean failed = !"Retour OK".equalsIgnoreCase(faultString) || !"0".equals(statusCode);

        info.put("StatusCode", res != null ? res.at("/Body/Simt101/SignatureResponseUrl/StatusCode").asText("") : "");
        info.put("ResultCode", res != null ? res.at("/Body/Simt101/SignatureResponseUrl/ResultCode").asText("") : "");
        info.put("status", failed ? "FAILED" : "PASSED");

        if (failed) {
            info.put("response.fault.string", faultString);
        }

        return info;
    }
}

File: desktop/src/job/onlinestandard-tpgroup-trace-extractor/src/onlinestandardtpgrouptraceextractor/trace_details_extractor.py
import logging
from io import StringIO
import xml.etree.ElementTree as ET

log = logging.getLogger(__name__)

def _local_name(tag: str) -> str:
    """Strip namespace from an XML tag."""
    return tag.rsplit("}", 1)[-1] if "}" in tag else tag

def extract_xml_value(xml: str, *path: str) -> str:
    """
    Stream-parse `xml` looking for nested elements matching `path`
    (e.g. extract_xml_value(res, "Body","Fault","Reason","Text","__text")).
    Returns first matching element text or "".
    """
    if not xml:
        return ""
    try:
        level = 0
        # iterparse over start/end keeps memory small
        for event, elem in ET.iterparse(StringIO(xml), events=("start","end")):
            if event == "start" and _local_name(elem.tag) == path[level]:
                level += 1
                if level == len(path):
                    # next event should carry text
                    text = (elem.text or "").strip()
                    return text
            elif event == "end" and level and _local_name(elem.tag) == path[level-1]:
                # popping back if we stepped into other branches
                level -= 1
        return ""
    except Exception as e:
        log.debug(f"StAX extraction failed for {'/'.join(path)}: {e}")
        return ""

def extract_abonnement(tags: dict) -> dict:
    info = {}
    req, res = tags.get("request.content",""), tags.get("response.content","")
    ackCode    = extract_xml_value(res, "Body","DemandeAbonPSResponse","Acquittement","AckCode")
    ackMessage = extract_xml_value(res, "Body","DemandeAbonPSResponse","Acquittement","AckMessage")
    typeDemande= extract_xml_value(req, "Body","DemandeAbonPSRequest","ModeAppel","TypeDemande")
    statusCode = tags.get("status.code","0")
    info.update({
        "TypeDemande": typeDemande,
        "AckCode":     ackCode,
        "AckMessage":  ackMessage,
        "request.num.ps": tags.get("request.num.ps")
    })
    failed = ackCode.upper()=="KO" or statusCode!="0"
    info["status"] = "FAILED" if failed else "PASSED"
    return info

def extract_simple(tags: dict) -> dict:
    info = {}
    res         = tags.get("response.content","")
    faultStr    = tags.get("response.fault.string","")
    statusCode  = tags.get("status.code","0")
    failed = faultStr.upper()!="RETOUR OK" or statusCode!="0"
    info["status"] = "FAILED" if failed else "PASSED"
    if failed:
        reason = extract_xml_value(res, "Body","Fault","Reason","Text","__text")
        if reason:
            info["failure.reason"]        = reason
            info["response.fault.string"] = faultStr
    info.update({
        "request.insee.beneficiaire": tags.get("request.insee.beneficiaire"),
        "request.num.amc":            tags.get("request.num.amc"),
        "request.num.ps":             tags.get("request.num.ps"),
        "request.uri":                tags.get("request.uri"),
    })
    return info

def extract_dmde_pratique(tags: dict) -> dict:
    info = {}
    res         = tags.get("response.content","")
    ackCode     = extract_xml_value(res, "Body","DonneesPSResponse","AckCode")
    ackMessage  = extract_xml_value(res, "Body","DonneesPSResponse","AckMessage")
    statusCode  = tags.get("status.code","0")
    failed = ackCode.upper()=="KO" or statusCode!="0"
    info.update({
        "request.num.ps": tags.get("request.num.ps"),
        "request.uri":    tags.get("request.uri"),
        "AckCode":        ackCode,
        "AckMessage":     ackMessage,
        "status":         "FAILED" if failed else "PASSED"
    })
    return info

def extract_convention(tags: dict) -> dict:
    info = {}
    faultString = tags.get("response.fault.string","")
    statusCode  = tags.get("status.code","0")
    failed = faultString.upper()!="RETOUR OK" or statusCode!="0"
    info.update({
        "request.num.amc": tags.get("request.num.amc"),
        "request.num.ps":  tags.get("request.num.ps"),
        "status":          "FAILED" if failed else "PASSED"
    })
    if failed:
        info["response.fault.string"] = faultString
    return info

def extract_restitution(tags: dict) -> dict:
    info = {}
    faultString = tags.get("response.fault.string","")
    faultCode   = tags.get("response.fault.code","")
    statusCode  = tags.get("status.code","0")
    failed = faultString.upper()!="RETOUR OK" or statusCode!="0"
    info["request.num.ps"] = tags.get("request.num.ps")
    info["status"]         = "FAILED" if failed else "PASSED"
    if failed:
        info["response.fault.code"]   = faultCode
        info["response.fault.string"] = faultString
    return info

def extract_inter_amc(tags: dict) -> dict:
    info = {}
    res         = tags.get("response.content","")
    codeRetour  = extract_xml_value(res,
                   "Body","searchResponse","searchResult","etat_demande","code_retour")
    message     = extract_xml_value(res,
                   "Body","searchResponse","searchResult","etat_demande","message")
    failed = codeRetour!="1"
    info["status"]      = "FAILED" if failed else "PASSED"
    if failed:
        info["code_retour"] = codeRetour
        info["message"]     = message
    return info

def extract_signature(tags: dict) -> dict:
    info = {}
    res         = tags.get("response.content","")
    faultString = tags.get("response.fault.string","")
    statusCode  = tags.get("status.code","0")
    status_failed = faultString.upper()!="RETOUR OK" or statusCode!="0"
    statusCodeOut  = extract_xml_value(res,
                     "Body","Simt101","SignatureResponseUrl","StatusCode")
    resultCodeOut  = extract_xml_value(res,
                     "Body","Simt101","SignatureResponseUrl","ResultCode")
    info.update({
        "StatusCode": statusCodeOut,
        "ResultCode": resultCodeOut,
        "status":     "FAILED" if status_failed else "PASSED"
    })
    if status_failed:
        info["response.fault.string"] = faultString
    return info

# Map of operation → extractor
_EXTRACTORS = {
    "AbonnementPS":   extract_abonnement,
    "IDB":            extract_simple,
    "CLC":            extract_simple,
    "DmdePratique":   extract_dmde_pratique,
    "ConventionPS":   extract_convention,
    "restitutionAMC": extract_restitution,
    "InterAMC":       extract_inter_amc,
    "DmdeSignature":  extract_signature,
}

def extract_details(operation: str, tags: dict) -> dict:
    """
    Dispatch to the right function by operation name.
    """
    try:
        return _EXTRACTORS.get(operation, lambda t: {})(tags)
    except Exception as e:
        log.warn(f"Failed to extract for {operation}: {e}")
        return {}